tiny_dnn/util/product.h

```diff
@@ -516,6 +551,61 @@ void fill(T *dst, size_t size, T value) {
 
 }  // namespace detail
 
+#ifdef CNN_USE_AVX
+// vertically accumulate 'n' AVX registers into single register.
+template <typename aligned>
+#ifdef CNN_USE_DOUBLE
+CNN_MUSTINLINE VECTORIZE_TYPE::register_type accumulate(const double *start,
+                                                        const size_t &nblocks) {
+#else
+CNN_MUSTINLINE VECTORIZE_TYPE::register_type accumulate(const float *start,
+                                                        const size_t &nblocks) {
+#endif
+  const size_t n4                  = nblocks / 4;
+  const size_t n2                  = (nblocks % 4) / 2;
+  const size_t n1                  = nblocks % 2;
+  VECTORIZE_TYPE::register_type v0 = VECTORIZE_TYPE::load<aligned>(start + 000);
+  VECTORIZE_TYPE::register_type v1 = VECTORIZE_TYPE::load<aligned>(start + 010);
+  VECTORIZE_TYPE::register_type v2 = VECTORIZE_TYPE::load<aligned>(start + 020);
+  VECTORIZE_TYPE::register_type v3 = VECTORIZE_TYPE::load<aligned>(start + 030);
```

Comment:
These 8x displacement indexes are for float path. So you need to use 4x ones for double path.